### DISCUSSION TITLE
Global middleware per Router?

### DISCUSSION THREAD

**Discussion by Vesafary** [aws-powertools/powertools-lambda-python#4551] [Q&A]
Let's say I have multiple Routers to split routes, is it possible to give each router its own global middleware?

Right now, in my test setup, it seems like they are all piled together:
```python
# Sketch setup

class TestMiddleware(BaseMiddlewareHandler):
    def __init__(self):
        super().__init__()

    def handler(self, app: APIGatewayRestResolver, next_middleware: NextMiddleware) -> Response:
        print("Hello from middleware")
        return next_middleware(app)


router = Router()
router.use(middlewares=[
    TestMiddleware(), 
])
router2 = Router()

app = APIGatewayRestResolver(debug=True)
app.include_router(router)
app.include_router(router2)

@router2.get(f"/test/")  # Would not expect any middleware being called here
def func():
    return "Ok"




# Processing Route:::func (/test/)
#
# Middleware Stack:
# =================
# TestMiddleware
# =================
```

**Comment by leandrodamascena** [comment]
Hey @Vesafary! Unfortunately, you cannot use individual middlewares per `Router` class because when you use `app.include_router()`, it merges all existing middlewares into the `app` instance and executes all of them. What you can have is a middleware per operation, like in this example: https://docs.powertools.aws.dev/lambda/python/latest/core/event_handler/api_gateway/#middleware.

I'm interested in understanding the use case for why you need middleware per Router. Perhaps you are organizing your routes based on a specific logic, such as domain/entities/others, and you would like to apply the same Middleware to all routes in this case?

**Comment by Vesafary** [comment]
Hi @leandrodamascena! Thanks for your answer.

There's a couple of reasons. The most important one to us: our application has a role-based authentication system (which is primarily handled by an API Gateway Authoriser), but a couple of endpoints have either role-based logic (e.g. an admin would fetch more (deleted) objects than a normal user) or instance-based logic (e.g. only the creator of an item is able to see it, or when the creator decides to share it). In these cases, a gateway authoriser can't do much more than accept the request, and the authentication has to be handled by the backend. This is what we're trying to do with a middleware, but to avoid someone forgetting to add it to a route (as per your example), or setting the wrong one, and to avoid code duplication and upon a refactor having to change dozens of endpoints, we want to add it globally as a "default" instead of on a per endpoint basis. However, that runs into the issue that some requests (e.g. OPTIONS, /swagger, some custom ones) have to be open-access. In my opinion, the nicest and easiest solution was to use multiple routers.

BTW, as a follow-up it would also be nice if we could nest routers. Right now, you can only include a router on the app-level with a prefix, but not on another one. Usecase for this would be even further separation and less repetition between multiple domains.
For instance, I could have a router per domain (/posts, /users, etc), and then have a router without middleware and one with, where I could do something like `middleware_router.use(middlewares=[...]).include_router(post_router).include_router(user_router)` and not have the middleware defined on both the post_router and user_router.

**Comment by Iamrodos** [comment]
I encountered a related challenge when refactoring a large Lambda handler by splitting routes into separate modules using the Router pattern.

  ### The Problem

When splitting routes into separate files, I needed to share middleware functions between the main app and router modules. However, this creates an import-time initialization problem:

  1. The main `lambda_web_function.py` defines middleware functions that reference the `app` instance
  2. A separate `lambda_web_admin.py` needs to use these same middlewares
  3. When Python imports the router module, `@router.route()` decorators execute immediately and need
  middleware function references
  4. But the actual middleware functions don't exist yet (circular import issue)

  ### Solution: Middleware Wrapper Pattern

  Solved this with a wrapper function pattern that handles the initialization timing:

  ```python
  # lambda_web_admin.py
  from aws_lambda_powertools.event_handler.api_gateway import Router

  # Placeholders for middlewares (set during initialization)
  _force_authenticated = None
  _setup_request = None
  _user_request_permission = None

  def init_admin_router(main_app, ..., force_auth, setup_req, user_perm):
      """Initialize the router with dependencies from main app."""
      global _force_authenticated, _setup_request, _user_request_permission
      _force_authenticated = force_auth
      _setup_request = setup_req
      _user_request_permission = user_perm

  # Wrapper functions that exist at import time and delegate at runtime
  def force_authenticated(app_param, next_middleware):
      """Wrapper for force_authenticated middleware."""
      return _force_authenticated(app_param, next_middleware)

  def setup_request(app_param, next_middleware):
      """Wrapper for setup_request middleware."""
      return _setup_request(app_param, next_middleware)

  def user_request_permission(app_param, next_middleware):
      """Wrapper for user_request_permission middleware."""
      return _user_request_permission(app_param, next_middleware)

  # Create Router and use wrappers in route decorators
  admin_router = Router()

  @admin_router.route(
      "/admin",
      middlewares=[force_authenticated, setup_request, user_request_permission],
      method=["GET", "POST"],
      compress=True,
  )
  def admin():
      # Route handler
      pass
```
  Then in the main file:
```python
  # lambda_web_function.py
  from lambda_web_admin import admin_router, init_admin_router

  # Initialize the router with actual middleware functions
  init_admin_router(
      main_app=app,
      force_auth=force_authenticated,
      setup_req=setup_request,
      user_perm=user_request_permission
  )

  # Register the router
  app.include_router(admin_router)
```
### Why This Works

  - The wrapper functions exist at import time, so decorators can capture them
  - At runtime, wrappers delegate to the real middleware functions (set during initialization)
  - Avoids circular imports
  - Keeps middleware logic DRY (defined once in main module)

### Suggestion for Powertools

It would be helpful if the documentation included a pattern for sharing middlewares across router modules, as this is a common need when refactoring monolithic handlers. A built-in way to "inject" middlewares into routers after import would also be cleaner than this workaround.

**Comment by leandrodamascena** [comment]
Hey @Vesafary and @Iamrodos! We recently merged `Depends()` support for type-safe dependency injection in route handlers (#8128), and this should address your use cases in a much cleaner way.

Instead of relying on global middlewares per Router (which get merged together), you can now declare dependencies per route using `Annotated[Type, Depends(...)]`. Each router can have its own set of dependencies without polluting other routers:

```python
from typing import Annotated
from aws_lambda_powertools.event_handler import APIGatewayRestResolver, Router, Depends

app = APIGatewayRestResolver()

# Admin router with its own auth dependency
admin_router = Router()

def require_admin() -> User:
    # validate admin role from request context
    ...

@admin_router.get("/admin/users")
def list_users(user: Annotated[User, Depends(require_admin)]):
    return {"users": [...]}

# Public router — no auth dependency needed
public_router = Router()

@public_router.get("/health")
def health():
    return {"status": "ok"}

app.include_router(admin_router)
app.include_router(public_router)
```

This also solves the circular import problem @Iamrodos described - dependencies are plain functions, no need for global state or wrappers.

What do you think about this? We are planning to release a new version in the next coming weeks.

**Comment by Iamrodos** [comment]
### TL;DR

`Depends()` solves the router middleware wiring problem well and is a clear win for new projects. For existing codebases built on `app.context` + middleware, the migration has three blockers:

1. **`Request` loses functionality** — `Depends()` functions can't access `current_event`, cookies, the actual request path, or any of the Powertools event helpers. This is a regression from what middleware receives today.
2. **`Depends()` functions can't access `app.context`** — making incremental migration impossible. You have to convert the entire middleware chain and all context references at once.
3. **Implicit ordering replaces explicit ordering** — a forgotten `Depends()` declaration silently breaks execution order rather than failing obviously.

---

Thanks @leandrodamascena for the `Depends()` work — this is a great direction. We're the project that posted the middleware wrapper pattern earlier in this thread, so we've been following this closely.

We've been evaluating what it would take to migrate from the current middleware + `app.context` pattern to `Depends()`, and wanted to share some practical feedback from that exercise.

### What works well

The core concept is sound. Dependencies as plain functions solve the circular import / init ceremony problem we described earlier. Being able to share dependency functions across router modules via normal imports — instead of the placeholder/wrapper/init pattern — is a significant improvement.

The dependency graph resolution (topological sorting, per-invocation caching) also handles our middleware chaining naturally. Our three middlewares have explicit ordering dependencies (auth → request parsing → permission check), and expressing those as `Depends()` parameters rather than a manually ordered list makes the relationships explicit.

### Concern: implicit ordering replaces explicit ordering

Today we control middleware order directly:

```python
middlewares=[force_authenticated, setup_request, user_request_permission]
```

With `Depends()`, order is inferred from the dependency graph. If a dependency function forgets to declare `Depends(require_auth)`, it might run before auth completes — a subtle bug rather than an obvious ordering mistake. This is a trade-off worth documenting clearly.

### Concern: `Request` is too thin for real middleware replacement

Our middleware functions currently receive the full `app` resolver, giving access to `app.current_event` with all the Powertools parsing and helpers. With `Depends()`, functions receive `Request`, which only exposes 7 properties: `route`, `path_parameters`, `method`, `headers`, `query_parameters`, `body`, `json_body`.

Things we lose moving from `app.current_event` to `Request`:

- **Actual request path** — `Request.route` is the pattern (`/admin/<path>`), not the path (`/admin/users`). We need the real path for login redirect state params and access-denied logging.
- **Event helpers** — `get_query_string_value(name, default)`, `get_header_value()` etc. Back to raw dict lookups.
- **Cookies** — `app.current_event.cookies` has no equivalent on `Request`.
- **The full event object** — we pass `app.current_event` directly to our authorization engine (Cedar policies). A `Depends()` function can't access it.
- **Request context** — Lambda/API Gateway metadata from `app.current_event.request_context`.

The whole value of Powertools is parsing and enriching the raw Lambda event. It feels like a step backwards to have `Depends()` hand us a reduced view of that.

**Suggestion:** Could `Depends()` functions receive `current_event` (or the resolver) instead of / in addition to `Request`? That would make `Depends()` a genuine middleware replacement rather than a limited alternative.

### Concern: migration path from `app.context`

Our codebase has ~550 references to `app.context` across handler functions — reading `user`, `form_data`, `handler_name`, `handler_action`, `handler_value`, `handler_mode` etc. that middleware placed there.

`Depends()` replaces this with return values passed as function parameters — cleaner in theory, but migrating 550 references is an all-or-nothing rewrite.

A practical bridge: inject `Depends()` results back into `app.context` at the top of each handler, so existing code continues to work:

```python
@router.route("/admin", method=["GET", "POST"])
def admin(
    user: Annotated[AuthUser, Depends(require_auth)],
    ctx: Annotated[RequestContext, Depends(setup_request)],
    perms: Annotated[PermissionResult, Depends(check_permission)],
    path="", subpath="",
):
    # Bridge: existing app.context references keep working
    app.append_context(user=user, form_data=ctx.form_data, ...)
    
    # ... 100+ lines of existing code unchanged
```

This works because `app` is accessible as a module global, but it would be even cleaner if `Depends()` had a first-class way to populate `app.context` — or if dependency functions could access `app.context` themselves, enabling one-middleware-at-a-time migration.

Happy to help test any of this if useful.

**Comment by Iamrodos** [comment]
Update: I raised these concerns in issue #8139. The maintainer is looking at enriching the `Request` object with `resolved_event` and `context` properties for `Depends()` functions, which would address the gaps discussed here. See that issue for the full conversation on how middleware and `Depends()` complement each other.

**Comment by Iamrodos** [comment]
@leandrodamascena following on from the conversation in #8139, I wanted to circle back to the original problem in this discussion. The middleware import problem that I described with the placeholder/wrapper pattern is still unsolved.

`Depends()` solves the data-sharing side (and the enriched `Request` with `resolved_event` and `context` will make that even better). But as you said, middleware is the right tool when you need the full resolver lifecycle (pre/post processing, short-circuiting, etc.). Those use cases must stay as middleware.

The core issue is that `@router.route()` decorators capture the `middlewares=` list at import time. When middleware functions are defined in the main handler module (because they depend on `app`, `auth`, config, etc.), router modules in separate files can't import them without circular dependencies. That's why I ended up with the placeholder/wrapper/init pattern across all my router modules.

`Depends()` solved this exact timing problem for dependencies by deferring resolution to request time. I wonder if a similar pattern could work for middleware. Something like a `Middleware()` wrapper that holds a reference but defers resolution:

```python
@admin_router.route(
    "/admin",
    middlewares=[Middleware(force_authenticated), Middleware(setup_request)],
)
```

The challenge is that this only helps if the function can actually be imported. If the circular dependency prevents the import entirely, lazy resolution at call time doesn't help because the import still fails at module load time. So it might need string references or some other late-binding mechanism, which gets ugly fast.

Another option would be passing middleware at `include_router` time, since that happens in handler.py where the middleware functions are available:

```python
app.include_router(admin_router, middlewares=[force_authenticated, setup_request])
```

But that has two drawbacks. It applies the same middleware to all routes in the router (my routes use different middleware combinations per route), and it separates the middleware declaration from the route definition, making it harder to understand what a route does by reading one file.

I don't have a clean solution here, but I wanted to flag that the per-router middleware import problem is still an open pain point for anyone splitting routes into multiple files. If the team has ideas on how to address this at the framework level, I'd be keen to help build it.

**Comment by Iamrodos** [comment]
@leandrodamascena I have spent more time this morning wrestling with the middleware-across-files problem, trying to understand it all better. Here is some evolved thinking I am interested in your thoughts on.

Part of the problem is self-inflicted. My middleware functions receive `app` as their first parameter, but the helper functions they call ignore that and read a module-level `app` global instead. That's what forces my middleware to live in handler.py, which creates the circular import, which forces the placeholder/wrapper/init ceremony in every router module.

I could fix this by refactoring those helpers to accept `app` as a parameter and extracting everything into shared modules (the Flask "extensions.py" pattern). No Powertools changes needed.

But looking at how other frameworks handle this same problem, there's a common pattern:

| Framework | Solution |
|---|---|
| Flask | `current_app` proxy via contextvars |
| FastAPI | `request.app` gives you the application |
| Express.js | `req.app` gives you the application |

They all provide a way to access the active application during request processing without importing it. This prevents developers from falling into the globals pattern in the first place.

For Powertools, this could be a `get_current_app()` function. The implementation is small because Powertools already sets up class-level state in `resolve()` and already uses `contextvars` in the logging module:

```python
# aws_lambda_powertools/event_handler/_context.py
from contextvars import ContextVar

_current_app_var: ContextVar = ContextVar("powertools_current_app")

def get_current_app():
    """Returns the active resolver during request processing."""
    try:
        return _current_app_var.get()
    except LookupError:
        raise RuntimeError(
            "get_current_app() is only available during request processing."
        )
```

Then two lines in `resolve()` to set/clear it. That's the entire framework change.

The value is that utility functions can access the resolver from anywhere without imports from the handler module:

```python
# shared/middleware_helpers.py (no circular imports)
from aws_lambda_powertools.event_handler import get_current_app

def get_session_user():
    app = get_current_app()
    return app.context.get("user")

def log_request_path():
    app = get_current_app()
    logger.info(f"Processing {app.current_event.path}")
```

This means middleware helpers can live in shared modules and be imported directly by router files. No placeholders, no wrappers, no init functions.

For documentation, I think this fits naturally in the "Sharing contextual data" section alongside `append_context`, framed as "accessing the resolver from utility functions when splitting routes into separate files."

This is obviously not a priority compared to the `Request` enrichment work on #8139. But given that the split-routes-with-middleware problem is what started this whole discussion, it felt worth raising. What do you think?

**Reply by leandrodamascena** [reply]
Hey @Iamrodos, the circular import problem you're hitting comes from helpers reading a module-level `app` global instead of using the `app` parameter that middleware already receives. Once you refactor those helpers to accept `app` as a parameter, they can live in shared modules and the circular dependency disappears. No framework changes needed I think.

On `get_current_app()` - it's a valid pattern (Flask does it), but I'd rather not add another way to access the resolver. We already have `app` as the middleware parameter, `BaseRouter.current_event` at class level, and soon `request.context` in `Depends()`. Adding `contextvars` on top of that makes the surface area harder to reason about.

What do you think? That make sense?

**Reply by Iamrodos** [reply]
@leandrodamascena I feel like I have spent the day cleaning out the gutters of accumulated technical debt, and you were right about where the blockage was. The extraction worked and the circular import problem disappeared.

What surprised me was how much else came along. Once shared helpers and singletons moved out of the main handler module, the init/wrapper pattern I had built for the routers had nothing left to inject and deleted itself. **Even `app` itself can be imported from the handler module by route files.** As Python supports partial imports, by the time the handler loads the route modules the names they need are already defined. "Explicit DI" turned into ordinary imports, which is what it should have been all along. The wiring I was hoping to fix at the framework level was scaffolding for a god-module I had never broken up.

My refactor a while ago into multiple route files was good but always felt a little messy and un-pythonic. Its now how it should be. Sweet.

**Comment by Vesafary** [comment]
Hey, this looks pretty promising. Just like @Iamrodos our codebase has grown significantly so a refactor will be a huge effort, but I will definitely check it out! Thanks for the effort on this :heart:

**Reply by Iamrodos** [reply]
@Vesafary, to follow up with some numbers in case it helps you weigh the effort.

The refactor came out smaller than I feared. Across 8 route files and a ~25k-line handler package, the work was about 1.5 days once v3.28.0 shipped. Lots of testing. Finished it this morning.

My baseline coupling from route code to the resolver was 748 references: 557 reads of app.context, 26 reads of app.current_event, and 165 calls to local helpers that were themselves wrappers around app.context["user"]. After migration every one of those is gone from route files. The only app.current_event reads left live in middleware.py and handler.py, which is the framework layer where they belong.

The sequencing that saved time was doing the extraction work before touching Depends():

1. Move singletons (database client, auth engine, cookie/crypto config) out of the handler module into a shared services.py.
2. Refactor middleware helpers to take app or request as a parameter instead of reading module globals.
3. Move the middleware functions themselves into a dedicated middleware.py.

At that point, router modules could import middleware by name. The placeholder/wrapper/init ceremony I described upthread deleted itself. No Depends() yet, no framework changes, and about 420 lines of boilerplate gone.

Then Depends() went on top. Four middleware became user_required, user_optional, parse_request, authorize. Two stayed as middleware because they need to short-circuit with responses (user_required issues a Cognito redirect for unauthenticated GETs; authorize issues HX-Redirect on permission denial). That split tracks @leandrodamascena's point: Depends for data injection, middleware for responses.

Route handlers dropped their app.context[...] reads and now declare current_user: Annotated[User, Depends(get_current_user)] and handler: Annotated[RequestContext, Depends(request_context)] as parameters. 

If you plan the same work: the biggest effort for me was not Depends() itself but the prior extraction. Once helpers take parameters instead of reaching for globals, the rest falls out naturally. It was the right refactor either way.
